### ISSUE TITLE
jme3test.post.TestSSAO2 throws OpenGLException

### ISSUE THREAD

**Issue by stephengold** [jMonkeyEngine/jmonkeyengine#995]
Currently (in `master` branch at 3.3-6689) with assertions enabled, the `TestSSAO2` app crashes with the following stack trace:
```
Dec 31, 2018 10:11:41 AM com.jme3.scene.plugins.ogre.MeshLoader load
WARNING: Cannot locate Models/Teapot/Teapot.material for model Models/Teapot/Teapot.mesh.xml
Dec 31, 2018 10:11:41 AM com.jme3.scene.plugins.ogre.MeshLoader applyMaterial
WARNING: Cannot locate Material for model Models/Teapot/Teapot.mesh.xml
----------------- SSAO UI Debugger --------------------
-- Sample Radius : press Y to increase, H to decrease
-- AO Intensity : press U to increase, J to decrease
-- AO scale : press I to increase, K to decrease
-- AO bias : press O to increase, P to decrease
-- Toggle AO on/off : press space bar
-- Use only AO : press Num pad 0
-- Output config declaration : press P
-------------------------------------------------------
Dec 31, 2018 10:11:42 AM com.jme3.app.LegacyApplication handleError
SEVERE: An OpenGL error has occured!
org.lwjgl.opengl.OpenGLException: Invalid operation (1282)
	at org.lwjgl.opengl.Util.checkGLError(Util.java:59)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.checkGLError(LwjglAbstractDisplay.java:136)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.runLoop(LwjglAbstractDisplay.java:157)
	at com.jme3.system.lwjgl.LwjglDisplay.runLoop(LwjglDisplay.java:197)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:232)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by stephengold** [comment]
I re-ran the test in current `master` branch with `setGraphicsDebug(true)` and got some clues as to what goes wrong:
```text
Apr 07, 2020 6:06:19 PM com.jme3.system.JmeDesktopSystem initialize
INFO: Running on jMonkeyEngine 3.4.0-SNAPSHOT
 * Branch: master
 * Git Hash: b8287a2
 * Build Date: 2020-04-07
Apr 07, 2020 6:06:19 PM com.jme3.system.lwjgl.LwjglContext printContextInitInfo
INFO: LWJGL 2.9.3 context running on thread jME3 Main
 * Graphics Adapter: null
 * Driver Version: null
 * Scaling Factor: 1
Apr 07, 2020 6:06:19 PM com.jme3.renderer.opengl.GLRenderer loadCapabilitiesCommon
INFO: OpenGL Renderer Information
 * Vendor: NVIDIA Corporation
 * Renderer: GeForce GT 545/PCIe/SSE2
 * OpenGL Version: 4.6.0 NVIDIA 390.116
 * GLSL Version: 4.60 NVIDIA
 * Profile: Compatibility
Apr 07, 2020 6:06:20 PM com.jme3.audio.openal.ALAudioRenderer initOpenAL
INFO: Audio Renderer Information
 * Device: OpenAL Soft
 * Vendor: OpenAL Community
 * Renderer: OpenAL Soft
 * Version: 1.1 ALSOFT 1.15.1
 * Supported channels: 64
 * ALC extensions: ALC_ENUMERATE_ALL_EXT ALC_ENUMERATION_EXT ALC_EXT_CAPTURE ALC_EXT_DEDICATED ALC_EXT_disconnect ALC_EXT_EFX ALC_EXT_thread_local_context ALC_SOFT_loopback
 * AL extensions: AL_EXT_ALAW AL_EXT_DOUBLE AL_EXT_EXPONENT_DISTANCE AL_EXT_FLOAT32 AL_EXT_IMA4 AL_EXT_LINEAR_DISTANCE AL_EXT_MCFORMATS AL_EXT_MULAW AL_EXT_MULAW_MCFORMATS AL_EXT_OFFSET AL_EXT_source_distance_model AL_LOKI_quadriphonic AL_SOFT_buffer_samples AL_SOFT_buffer_sub_data AL_SOFTX_deferred_updates AL_SOFT_direct_channels AL_SOFT_loop_points AL_SOFT_source_latency
Apr 07, 2020 6:06:20 PM com.jme3.audio.openal.ALAudioRenderer initOpenAL
WARNING: Pausing audio device not supported.
Apr 07, 2020 6:06:20 PM com.jme3.audio.openal.ALAudioRenderer initOpenAL
INFO: Audio effect extension version: 1.0
Apr 07, 2020 6:06:20 PM com.jme3.audio.openal.ALAudioRenderer initOpenAL
INFO: Audio max auxiliary sends: 4
Apr 07, 2020 6:06:20 PM com.jme3.scene.plugins.ogre.MeshLoader load
WARNING: Cannot locate Models/Teapot/Teapot.material for model Models/Teapot/Teapot.mesh.xml
Apr 07, 2020 6:06:20 PM com.jme3.scene.plugins.ogre.MeshLoader applyMaterial
WARNING: Cannot locate Material for model Models/Teapot/Teapot.mesh.xml
----------------- SSAO UI Debugger --------------------
```
 [... many messages omitted ...]
```text
[JME3] OpenGL debug message
       ID: 131185
       Source: API
       Type: OTHER
       Severity: null
       Message: Buffer detailed info: Buffer object 23 (bound to GL_ELEMENT_ARRAY_BUFFER_ARB, usage hint is GL_DYNAMIC_DRAW) will use VIDEO memory as the source for buffer object operations.
java.lang.Exception: Stack trace
	at java.lang.Thread.dumpStack(Thread.java:1336)
	at com.jme3.system.lwjgl.LwjglGLDebugOutputHandler.handleMessage(LwjglGLDebugOutputHandler.java:76)
	at org.lwjgl.opengl.GL15.nglBufferData(Native Method)
	at org.lwjgl.opengl.GL15.glBufferData(GL15.java:170)
	at com.jme3.renderer.lwjgl.LwjglGL.glBufferData(LwjglGL.java:92)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.jme3.renderer.opengl.GLDebug.invoke(GLDebug.java:107)
	at com.sun.proxy.$Proxy0.glBufferData(Unknown Source)
	at com.jme3.renderer.opengl.GLRenderer.updateBufferData(GLRenderer.java:2712)
	at com.jme3.renderer.opengl.GLRenderer.drawTriangleList(GLRenderer.java:2965)
	at com.jme3.renderer.opengl.GLRenderer.renderMeshDefault(GLRenderer.java:3172)
	at com.jme3.renderer.opengl.GLRenderer.renderMesh(GLRenderer.java:3200)
	at com.jme3.material.logic.DefaultTechniqueDefLogic.renderMeshFromGeometry(DefaultTechniqueDefLogic.java:70)
	at com.jme3.material.logic.DefaultTechniqueDefLogic.render(DefaultTechniqueDefLogic.java:95)
	at com.jme3.material.Technique.render(Technique.java:166)
	at com.jme3.material.Material.render(Material.java:1028)
	at com.jme3.renderer.RenderManager.renderGeometry(RenderManager.java:614)
	at com.jme3.renderer.queue.RenderQueue.renderGeometryList(RenderQueue.java:266)
	at com.jme3.renderer.queue.RenderQueue.renderQueue(RenderQueue.java:302)
	at com.jme3.renderer.RenderManager.renderViewPortQueues(RenderManager.java:905)
	at com.jme3.renderer.RenderManager.flushQueue(RenderManager.java:779)
	at com.jme3.renderer.RenderManager.renderViewPort(RenderManager.java:1108)
	at com.jme3.renderer.RenderManager.render(RenderManager.java:1166)
	at com.jme3.app.SimpleApplication.update(SimpleApplication.java:273)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.runLoop(LwjglAbstractDisplay.java:157)
	at com.jme3.system.lwjgl.LwjglDisplay.runLoop(LwjglDisplay.java:196)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:239)
	at java.lang.Thread.run(Thread.java:748)
[JME3] OpenGL debug message
       ID: 1282
       Source: API
       Type: ERROR
       Severity: HIGH
       Message: GL_INVALID_OPERATION error generated. The query is active to another target.
java.lang.Exception: Stack trace
	at java.lang.Thread.dumpStack(Thread.java:1336)
	at com.jme3.system.lwjgl.LwjglGLDebugOutputHandler.handleMessage(LwjglGLDebugOutputHandler.java:76)
	at org.lwjgl.opengl.GL15.nglGetQueryObjectiv(Native Method)
	at org.lwjgl.opengl.GL15.glGetQueryObjecti(GL15.java:464)
	at com.jme3.renderer.lwjgl.LwjglGL.glGetQueryObjectiv(LwjglGL.java:552)
	at sun.reflect.GeneratedMethodAccessor26.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.jme3.renderer.opengl.GLDebug.invoke(GLDebug.java:107)
	at com.sun.proxy.$Proxy0.glGetQueryObjectiv(Unknown Source)
	at com.jme3.renderer.opengl.GLRenderer.isTaskResultAvailable(GLRenderer.java:3262)
	at com.jme3.app.DetailedProfiler.appStep(DetailedProfiler.java:60)
	at com.jme3.app.SimpleApplication.update(SimpleApplication.java:245)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.runLoop(LwjglAbstractDisplay.java:157)
	at com.jme3.system.lwjgl.LwjglDisplay.runLoop(LwjglDisplay.java:196)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:239)
	at java.lang.Thread.run(Thread.java:748)
Apr 07, 2020 6:06:20 PM com.jme3.app.LegacyApplication handleError
SEVERE: Uncaught exception thrown in Thread[jME3 Main,5,main]
com.jme3.renderer.RendererException: An OpenGL error occurred - Invalid operation (Error Code: 1282)
	at com.jme3.renderer.opengl.GLDebug.checkError(GLDebug.java:101)
	at com.jme3.renderer.opengl.GLDebug.invoke(GLDebug.java:113)
	at com.sun.proxy.$Proxy0.glGetQueryObjectiv(Unknown Source)
	at com.jme3.renderer.opengl.GLRenderer.isTaskResultAvailable(GLRenderer.java:3262)
	at com.jme3.app.DetailedProfiler.appStep(DetailedProfiler.java:60)
	at com.jme3.app.SimpleApplication.update(SimpleApplication.java:245)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.runLoop(LwjglAbstractDisplay.java:157)
	at com.jme3.system.lwjgl.LwjglDisplay.runLoop(LwjglDisplay.java:196)
	at com.jme3.system.lwjgl.LwjglAbstractDisplay.run(LwjglAbstractDisplay.java:239)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by MeFisto94** [comment]
Can you maybe try to remove the Profiler AppState and see if that "solves" it?

**Comment by stephengold** [comment]
Good idea! Commenting out the `ProfilerAppState` seems to avoid the issue.

**Comment by MeFisto94** [comment]
Actually not an idea, but:
```
com.jme3.renderer.RendererException: An OpenGL error occurred - Invalid operation (Error Code: 1282)
	at com.jme3.renderer.opengl.GLDebug.checkError(GLDebug.java:101)
	at com.jme3.renderer.opengl.GLDebug.invoke(GLDebug.java:113)
	at com.sun.proxy.$Proxy0.glGetQueryObjectiv(Unknown Source)
	at com.jme3.renderer.opengl.GLRenderer.isTaskResultAvailable(GLRenderer.java:3262)
	at com.jme3.app.DetailedProfiler.appStep(DetailedProfiler.java:60)
	at com.jme3.app.SimpleApplication.update(SimpleApplication.java:245)
```

Now we need to find out what isTaskResultAvailable does with glGetQueryObjecti and why that _query_ is active to another target.

Also seeing that reminds me we definitely need the logging callback for lwjgl3 as well

**Comment by MeFisto94** [comment]
Turns out I have no clue and we'd need to defer this to @riccardobl 
https://github.com/jMonkeyEngine/jmonkeyengine/blob/master/jme3-core/src/main/java/com/jme3/renderer/opengl/GLRenderer.java#L3261

Looks like it's only really related to the profiling mechanisms used, though. Not sure why they conflict with SSAO

**Comment by stephengold** [comment]
OK. I'm just glad to see some progress on this issue.

**Comment by riccardobl** [comment]
The issue is the last query not being ended before trying to collect the results.
It should be fixed in https://github.com/jMonkeyEngine/jmonkeyengine/commit/076279db87d415f97caf37679fac8b2f1b484088

**Comment by stephengold** [comment]
Thank you @riccardobl . I tested that fix and it looks good.

**Comment by stephengold** [comment]
Fixed in v3.3 branch at 6d2ea28 .

**Comment by stephengold** [comment]
This issue appears to have been a duplicate of #794 .
